sys/exec.c

```diff
@@ -0,0 +1,292 @@
+#include <common.h>
+#include <exec.h>
+#include <stdc.h>
+#include <elf/mips_elf.h>
+#include <vm_map.h>
+#include <vm_pager.h>
+#include <thread.h>
+#include <string.h>
+#include <errno.h>
+
+extern uint8_t _binary_prog_uelf_start[];
+extern uint8_t _binary_prog_uelf_size[];
+extern uint8_t _binary_prog_uelf_end[];
+
+int get_elf_image(const exec_args_t *args, uint8_t **out_image,
+                  size_t *out_size) {
+  if (strcmp(args->prog_name, "prog") == 0) {
+    *out_image = _binary_prog_uelf_start;
+    *out_size = (size_t)_binary_prog_uelf_size;
+    return 0;
+  } else {
+    return -1;
+  }
+}
+
+// Places program args onto the stack.
+// Also modifies value pointed by stack_bottom_p to reflect on changed
+// stack bottom address.
+// The stack layout will be as follows:
+//
+//  ----------- stack segment high address
+//  | argv[n] |
+//  |   ...   |  each of argv[i] is a null-terminated string
+//  | argv[1] |
+//  | argv[0] |
+//  |---------|
+//  | argv    |  the argument vector storing pointers to argv[0..n]
+//  |---------|
+//  | argc    |  a single uint32 declaring the number of arguments (n)
+//  |---------|
+//  | program |
+//  |  stack  |
+//  |   ||    |
+//  |   \/    |
+//  |         |
+//  |   ...   |
+//  ----------- stack segment low address
+//
+//
+// After this function runs, the value pointed by stack_bottom_p will
+// be the address where argc is stored, which is also the bottom of
+// the now empty program stack, so that it can naturally grow
+// downwards.
+//
+static void prepare_program_stack(const exec_args_t *args,
+                                  vm_addr_t *stack_bottom_p) {
+  vm_addr_t stack_end = *stack_bottom_p;
+  // Begin by calculting arguments total size. This has to be done
+  // in advance, because stack grows downwards.
+  size_t total_arg_size = 0;
+  for (size_t i = 0; i < args->argc; i++) {
+    total_arg_size += strlen(args->argv[i]) + 1;
+  }
+  // Store arguments, creating the argument vector.
+  vm_addr_t arg_vector[args->argc];
+  vm_addr_t p = *stack_bottom_p - total_arg_size;
+  for (int i = 0; i < args->argc; i++) {
+    size_t n = strlen(args->argv[i]) + 1;
+    arg_vector[i] = p;
+    memcpy((uint8_t *)p, args->argv[i], n);
+    p += n;
+  }
+  assert(p == stack_end);
+  // Move the stack down and word-align it downwards
+  *stack_bottom_p = (*stack_bottom_p - total_arg_size) & 0xfffffffc;
```

Comment:
Please read [MIPS ABIs Described – Calling Conventions](https://dmz-portal.mips.com/mw/images/f/fe/MD00305-2B-ABIDESC-SPC-01.03.pdf).

I think that strings should be 4 bytes aligned, and top of the stack - 8 bytes aligned.